pythainlp/util/wordtonum.py

```diff
@@ -2,20 +2,29 @@
 """
 Convert number in words to a computable number value
 
-Adapted from Korakot Chaovavanich's notebook
+First version of the code adapted from Korakot Chaovavanich's notebook
 https://colab.research.google.com/drive/148WNIeclf0kOU6QxKd6pcfwpSs8l-VKD#scrollTo=EuVDd0nNuI8Q
 """
 import re
-from typing import Iterable, List
 
 from pythainlp.tokenize import Tokenizer
 
-_THAIWORD_NUMS = set("ศูนย์ หนึ่ง เอ็ด สอง ยี่ สาม สี่ ห้า หก เจ็ด แปด เก้า".split())
-_THAIWORD_UNITS = set("สิบ ร้อย พัน หมื่น แสน ล้าน".split())
-_THAIWORD_NUMS_UNITS = _THAIWORD_NUMS | _THAIWORD_UNITS
-
-_THAI_INT_MAP = {
-    "ศูนย์": 0,
+_re_thai_numerals = re.compile(
+    r"(((|หนึ่ง|สอง|สาม|สี่|ห้า|หก|เจ็ด|แปด|เก้า)แสน)?"
+    r"((|หนึ่ง|สอง|สาม|สี่|ห้า|หก|เจ็ด|แปด|เก้า)หมื่น)?"
+    r"((|หนึ่ง|สอง|สาม|สี่|ห้า|หก|เจ็ด|แปด|เก้า)พัน)?"
+    r"((|หนึ่ง|สอง|สาม|สี่|ห้า|หก|เจ็ด|แปด|เก้า)ร้อย)?"
+    r"((|หนึ่ง|ยี่|สาม|สี่|ห้า|หก|เจ็ด|แปด|เก้า)สิบ)?"
+    r"(|หนึ่ง|เอ็ด|สอง|สาม|สี่|ห้า|หก|เจ็ด|แปด|เก้า)?ล้าน)*"
+    r"(((|หนึ่ง|สอง|สาม|สี่|ห้า|หก|เจ็ด|แปด|เก้า)แสน)?"
+    r"((|หนึ่ง|สอง|สาม|สี่|ห้า|หก|เจ็ด|แปด|เก้า)หมื่น)?"
+    r"((|หนึ่ง|สอง|สาม|สี่|ห้า|หก|เจ็ด|แปด|เก้า)พัน)?"
+    r"((|หนึ่ง|สอง|สาม|สี่|ห้า|หก|เจ็ด|แปด|เก้า)ร้อย)?"
+    r"((|หนึ่ง|ยี่|สาม|สี่|ห้า|หก|เจ็ด|แปด|เก้า)สิบ)?"
+    r"(|หนึ่ง|เอ็ด|สอง|สาม|สี่|ห้า|หก|เจ็ด|แปด|เก้า)?)"
+)
```

Comment:
it would be good if we have some explanation here.

Reply:
It is the full expansion of regexp roughly equivalent to the following EBNF grammar:
```ebnf
thaiword = { six_figures, "ล้าน" }, six_figures;
six_figures = [ [ digit ], "แสน" ], [ [ digit ], "หมื่น" ], [ [ digit ], "พัน" ], [ [ digit ], "ร้อย" ], [ [ digit ], "สิบ" ],  [ digit ];
digit = "หนึ่ง" | "สอง" | "สาม" | "สี่" | "ห้า" | "หก" | "เจ็ด" | "แปด" | "เก้า";
```
Would it be okay to put this as a comment in the code? Or would you prefer to have the code refactored to reflect this grammar pattern?

Reply:
New code was pushed to explicitly reflect this grammar.